File: explorer-backend/src/Models/API/BlockResponse.cs
using ExplorerBackend.Models.Data;
using ExplorerBackend.VeilStructs;

namespace ExplorerBackend.Models.API;

public class BlockResponse
{
    public bool Found { get; set; }
    public BlockBasicData? NextBlock { get; set; }
    public BlockBasicData? PrevBlock { get; set; }
    public string? VersionHex { get; set; }
    public int TxnCount { get; set; }
    public Block? Block { get; set; }
    public List<TransactionSimpleDecoded>? Transactions { get; set; }
}

public class BlockBasicData
{
    public string? Hash { get; set; }
    public int Height { get; set; }
}

public class TransactionSimpleDecoded
{
    public string? TxId { get; set; }
    public List<TxVinSimpleDecoded>? Inputs { get; set; }
    public List<TxVoutSimpleDecoded>? Outputs { get; set; }
    public bool IsBasecoin { get; set; }
    public bool IsCoinStake { get; set; }
    public bool IsZerocoinMint { get; set; }
    public bool IsZerocoinSpend { get; set; }
}

public class TxVinSimpleDecoded
{
    public string? PrevOutTx { get; set; }
    public uint PrevOutNum { get; set; }
    public List<string>? PrevOutAddresses { get; set; }
    public long PrevOutAmount { get; set; }

    public TxInType Type { get; set; }
    public long ZerocoinSpend { get; set; }

    public List<RingCTInput>? AnonInputs { get; set; }
}

public class TxVoutSimpleDecoded
{
    public List<string>? Addresses { get; set; }
    public bool IsOpreturn { get; set; }
    public bool IsCoinBase { get; set; }
    public long Amount { get; set; }
    public OutputTypes Type { get; set; }
    public txnouttype ScriptPubKeyType { get; set; }
    public ulong? CTFee { get; set; }

}

public class RingCTInput
{
    public string? TxId { get; set; }
    public uint VoutN { get; set; }
}

public enum TxInType : int
{
    DEFAULT = 0,
    ZEROCOIN_SPEND = 1,
    ANON = 2
}

File: explorer-frontend/src/models/API/BlockResponse.ts
import type { Block } from "@/models/Data/Block";

export interface TxVinSimpleDecoded {
    prevOutTx: string | null;
    prevOutNum: number;
    prevOutAddresses: Array<string> | null;
    prevOutAmount: number;

    type: TxInType;
    zerocoinSpend: number;

    anonInputs: Array<RingCTInput> | null;
}

export interface TxVoutSimpleDecoded {
    addresses: Array<string> | null;
    isOpreturn: boolean;
    isCoinBase: boolean;
    amount: number;
    type: OutputTypes;
    scriptPubKeyType: txnouttype;
    cTFee: number | null;
}

export interface RingCTInput {
    txId: string | null;
    voutN: number;
}

export enum TxInType {
    DEFAULT = 0,
    ZEROCOIN_SPEND = 1,
    ANON = 2
}

export enum OutputTypes {
    OUTPUT_NULL = 0, // marker for CCoinsView (0.14)
    OUTPUT_STANDARD = 1,
    OUTPUT_CT = 2,
    OUTPUT_RINGCT = 3,
    OUTPUT_DATA = 4,
}

export enum txnouttype {
    TX_NONSTANDARD,
    // 'standard' transaction types:
    TX_PUBKEY,
    TX_PUBKEYHASH,
    TX_SCRIPTHASH,
    TX_MULTISIG,
    TX_NULL_DATA, //!< unspendable OP_RETURN script that carries data
    TX_WITNESS_V0_SCRIPTHASH,
    TX_WITNESS_V0_KEYHASH,
    TX_WITNESS_UNKNOWN, //!< Only for Witness versions not already defined above

    TX_SCRIPTHASH256,
    TX_PUBKEYHASH256,
    TX_TIMELOCKED_SCRIPTHASH,
    TX_TIMELOCKED_SCRIPTHASH256,
    TX_TIMELOCKED_PUBKEYHASH,
    TX_TIMELOCKED_PUBKEYHASH256,
    TX_TIMELOCKED_MULTISIG,
    TX_ZEROCOINMINT,
}

export interface TransactionSimpleDecoded {
    txId: string | null;
    inputs: Array<TxVinSimpleDecoded> | null;
    outputs: Array<TxVoutSimpleDecoded> | null;
    isBasecoin: boolean;
    isCoinStake: boolean;
    isZerocoinMint: boolean;
    isZerocoinSpend: boolean;
}

export interface BlockBasicData {
    hash: string | null;
    height: number;
}

export interface BlockResponse {
    found: boolean;
    nextBlock: BlockBasicData;
    prevBlock: BlockBasicData;
    versionHex: string | null;
    txnCount: number;
    block: Block;
    transactions: Array<TransactionSimpleDecoded>;
}